viewing_party/party.py

```diff
@@ -1,23 +1,196 @@
+    #add user movie to list of dictionaries if user's watched title not in the list of friend watched titles
+    for user_movie in user_data['watched']:
+        if user_movie not in friends_watched:
+            unique_watched_user.append(user_movie)
+    
+    return unique_watched_user
+
+
+def get_friends_unique_watched(user_data):
+    user_watched = []
+    friends_unique_movies = []
+    # get the list of movies user watched
+    for movie in user_data["watched"]:
+        user_watched.append(movie)
+
+    # get the list of movies friends watched but user not
+    # outer loop goes through list of friends of user
+    for friend in user_data["friends"]:
+        # inner loop goes through list of friend's watched movies
+        for friend_movie in friend["watched"]:
+            if friend_movie not in user_watched:
+                if friend_movie not in friends_unique_movies:
+                    friends_unique_movies.append(friend_movie)
+
+    return friends_unique_movies
+
+
         
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+def get_available_recs(user_data):
+    friend_watched_not_empty = False # use flag. If flag becomes True, means at least 1 friend has watched movie
+    if user_data["watched"] == []: #if user_data "watched" is empty, return empty list []
+        return []
+    recommendations = []
+    # outer loop goes through list of friends of user
+    for friend in user_data["friends"]:
+        if len(friend["watched"]) > 0: # check friend's "watched" is not empty
+                friend_watched_not_empty = True
+        # inner loop goes through list of friend's watched movies
+        for friend_movie in friend["watched"]:
+            if friend_movie not in user_data["watched"] and friend_movie["host"] in user_data["subscriptions"]:
+                recommendations.append(friend_movie)
+
+    if friend_watched_not_empty == False:  #if friends "watched" is empty, recommendations is empty
+        recommendations = []
+    return recommendations
 
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
+def get_new_rec_by_genre(user_data):
+    # get the user's most frequently watched genre
+    # get list of recommended movies
+    # get_available_recs() function will return the list of recommended movies user didn't watched yet and at least on of friend has watched movie
+    # remove from the list of recommended movies the movie which genre is not most watched user genre
+    if user_data["watched"] == []:
+        return []
+    user_most_watched_genre = get_most_watched_genre(user_data)
+    new_rec_movies = get_available_recs(user_data)
+    for movie in new_rec_movies:
+        if movie["genre"] != user_most_watched_genre:
+            new_rec_movies.remove(movie)
+
+    return new_rec_movies

```

Comment:
Here's where we need to be careful about mutating a list when you are iterating through it. Instead of removing movies, let's define a new empty list and then append the movies we want.